kong/dynamic_hook/wrap_function_gen.lua

```diff
@@ -0,0 +1,205 @@
+local TEMPLATE = [[
+  return function(always_enabled_groups, group_name, original_func, handlers)
+    return function(%s)
+      if not always_enabled_groups[group_name] then
+        local dynamic_hook = ngx.ctx.dynamic_hook
+        if not dynamic_hook then
+          return original_func(%s)
+        end
+  
+        local enabled_groups = dynamic_hook.enabled_groups
+        if not enabled_groups[group_name] then
+          return original_func(%s)
+        end
+      end
+  
+      if handlers.before_mut then
+        local ok
+        ok, %s = pcall(handlers.before_mut, %s)
+        if not ok then
+          ngx.log(ngx.WARN,
+                  string.format("failed to run before_mut hook of %%s: %%s",
+                              group_name, a0))
+        end
+      end
+  
+      if handlers.befores then
+        for _, func in ipairs(handlers.befores) do
+          local ok, err = pcall(func, %s)
+          if not ok then
+            ngx.log(ngx.WARN,
+                    string.format("failed to run before hook of %%s: %%s",
+                                  group_name, err))
+          end
+        end
+      end
+      
+      ngx.log(ngx.ERR, debug.traceback())
```

Comment:
debug log for developing? can be removed or commented out

Reply:
done